colossalai/pipeline/stage_manager.py

```diff
@@ -32,6 +34,9 @@ def __init__(self, pg_mesh: ProcessGroupMesh, pipeline_axis: int, is_virtual: bo
         # the next rank of the last rank is rank0
         next_coord = coord[: self.pipeline_axis] + (coord[self.pipeline_axis] + 1,) + coord[self.pipeline_axis + 1 :]
         self.next_rank = self.pg_mesh.ravel(next_coord, self.pg_mesh.shape, mode="wrap")
+        # number of layer chunks in each stage for interleaved pipeline, with each device has discontinuous layers
+        self.num_model_chunks = num_model_chunks
```

Comment:
Can you keep these attributes in schedule?

Reply:
It seems to me that stage manager should have model chunk id information. As interleave added, it is only the last stage if the device is the last device _AND model chunk is the last chunk_. Notably, we now make a difference between last device and last stage https://github.com/hpcaitech/ColossalAI/blob/ad33868e8b8d9359b1b7eca20da27583fa2d6f32/colossalai/pipeline/stage_manager.py#L73